Gems/GraphModel/Code/Source/Model/Slot.cpp

```diff
@@ -601,71 +606,65 @@ namespace GraphModel
         return m_subId;
     }
 
-    const DataTypeList& Slot::GetPossibleDataTypes() const
-    {
-        // TODO: For now this will just return all the supported types, but eventually
-        // it return the subset of possible data types given the current configuration
-        // of the node.
-        return GetSupportedDataTypes();
-    }
-
     DataTypePtr Slot::GetDataType() const
     {
-        // If the slot definition only has a single data type, then that is returned.
-        // Otherwise, we can ask our parent node to find out what the active type is.
-        DataTypeList possibleDataTypes = GetPossibleDataTypes();
-        size_t numPossibleDataTypes = possibleDataTypes.size();
-        if (numPossibleDataTypes == 1)
-        {
-            return possibleDataTypes[0];
-        }
-        else if (numPossibleDataTypes > 1)
-        {
-            return GetParentNode()->GetDataType(shared_from_this());
-        }
-
-        return nullptr;
+        // Because some slots support multiple data types search for the one corresponding to the value
+        return GetDataTypeForValue(m_value);
     }
 
     void Slot::SetValue(const AZStd::any& value)
     {
         if (SupportsValue())
         {
 #if defined(AZ_ENABLE_TRACING)
-            DataTypePtr dataType = GetGraphContext()->GetDataTypeForValue(value);
-            AssertTypeMatch(dataType, "Slot::SetValue used with the wrong type");
+            DataTypePtr dataTypeUsed = GetDataTypeForValue(value);
+            AssertWithTypeInfo(SupportsValue(), dataTypeUsed, "This slot type does not support values");
```

Comment:
This assert is re-checking the same as the "if" above, can never fail. Does this need to be moved out?

Reply:
I did not even consider it. Was just trying to clean up the asserts. It could be moved out or we could get rid of the if.

Reply:
https://github.com/o3de/o3de/pull/10390